### ISSUE TITLE
Asset helper: audio

### ISSUE THREAD

**Issue by jodosha** [lotus/lotus#395]
Implement `audio` helper.

See [`Lotus::Assets::Helpers`](https://github.com/lotus/assets/blob/master/lib/lotus/assets/helpers.rb) and [`Lotus::Helpers::HtmlHelper`](https://github.com/lotus/helpers/blob/master/lib/lotus/helpers/html_helper.rb) for reference.

### Basic Usage

```erb
<%= audio 'song.ogg' %>
```

```html
<audio src="/assets/song.ogg"></audio>
```

### HTML Attributes

It must accept optional HTML attributes.

```erb
<%= audio 'song.ogg', autoplay: true, controls: true %>
```

```html
<audio src="/assets/song.ogg" autoplay controls></audio>
```

### Fallback Content

It accepts an optional block for fallback content for browsers not supporting `<audio>` tag.

```erb
<%=
  audio 'song.ogg' do
    "Your browser does not support the audio tag"
  end
%>
```

```html
<audio src="/assets/song.ogg">
  Your browser does not support the audio tag
</audio>
```

### Tracks

It should support `<track>` tag.

```erb
<%=
  audio 'song.ogg' do
    track(kind: 'captions', src: asset_path('song.en.vtt'), srclang: 'en', label: 'English')
    track(kind: 'captions', src: asset_path('song.it.vtt'), srclang: 'it', label: 'Italiano')
  end
%>
```

```html
<audio src="song.ogg">
  <track kind="captions" src="song.en.vtt" srclang="en" label="English">
  <track kind="captions" src="song.it.vtt" srclang="it" label="Italiano">
</audio>
```

A few notes:

  * For `<audio>` reference, see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio
  * In order to avoid concatenation between `track` invocations, use [`Lotus::Helpers::HtmlHelper`](https://github.com/lotus/helpers/blob/master/lib/lotus/helpers/html_helper.rb) internally.
  * The helper used in this example `asset_path`, isn't implemented yet. See https://github.com/lotus/lotus/issues/394

### Sources

It should support `<source>` tag.

```erb
<%=
  audio do
    text "Your browser does not support the audio tag"
    source src: asset_path('song.ogg'), type: 'audio/ogg'
    source src: asset_path('song.wav'), type: 'audio/wav'
  end
%>
```

```html
<audio>
  Your browser does not support the audio tag
  <source src="/assets/song.ogg" type="audio/ogg">
  <source src="/assets/song.wav" type="audio/wav">
</audio>
```

A few notes:

  * For `<audio>` reference, see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio
  * In order to avoid concatenation between `source` invocations, use [`Lotus::Helpers::HtmlHelper`](https://github.com/lotus/helpers/blob/master/lib/lotus/helpers/html_helper.rb) internally.
  * The helper `text` doesn't map any HTML tag, but it's useful to concat pure text. See https://github.com/lotus/helpers/pull/30
  * The helper used in this example `asset_path`, isn't implemented yet. See https://github.com/lotus/lotus/issues/394

### URL Prefix

This should take account of the [`prefix`](https://github.com/lotus/assets/blob/master/lib/lotus/assets/configuration.rb#L45) defined in `Lotus::Assets::Configuration`.

That means, if an application has a prefix: `/admin`, when one this helper is used, the prefix should be used.

```erb
<%= audio 'song.ogg' %>
```

```html
<audio src="/assets/admin/song.ogg"></audio>
```

### Digest Mode

If "Digest Mode" is enabled (usually in production), it should generate the file name with the digest suffix (see https://github.com/lotus/assets/blob/master/lib/lotus/assets/configuration.rb#L37).

```erb
<%= audio 'song.ogg' %>
```

```html
<audio src="/assets/song-78a21a377046e9ddb3bd9b509c22d4c3.ogg"></audio>
```

## Testing Notes

Add an integration test with Lotus::View, in order to make sure that the output isn't escaped twice and produces a valid HTML.

**Comment by klebervirgilio** [comment]
Can I work on this one? Will be my first contribution.

**Comment by joneslee85** [comment]
@klebervirgilio please go for it

**Comment by klebervirgilio** [comment]
:+1:

**Comment by klebervirgilio** [comment]
@joneslee85 @jodosha quick question...

The audio tag has these two boolean attributes: _autoplay_ and _controls_

According to the [W3C](http://www.w3.org/html/wg/drafts/html/master/infrastructure.html#boolean-attribute) we have two ways to represent/render them in the tag, either with a empty value or with the canonical name as a value.

Looks like the way we decided to go is  `booleantag="booleantag"`. [see](https://github.com/lotus/helpers/blob/master/lib/lotus/helpers/html_helper/empty_html_node.rb#L83)

I'm wondering if this:
```erb
<%= audio 'song.ogg', autoplay: true, controls: true %>
```
Could render this:
```html
<audio src="/assets/song.ogg" autoplay="autoplay" controls="controls"></audio>
```

IMHO, it's fine.
What're your thoughts guys?

Thanks!

**Comment by jodosha** [comment]
@klebervirgilio Hey, yes that is fine. Please use `Lotus::Helpers` to generate the HTML, so we can delegate this decision to that component.

**Comment by joneslee85** [comment]
@klebervirgilio gentle ping, how are you going with the PR? Do you need any help?

**Comment by klebervirgilio** [comment]
Hi @joneslee85, thanks for reaching out.

I've done most of the work already, you can check it out [here](https://github.com/klebervirgilio/helpers/commit/2e2e8ad5fe8365a04bdbf426edfba1152ecc750b).

My initial work were basically this  humble [module](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/lib/lotus/helpers/audio_helper.rb#L4-L45) for which I've written these [specs](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/test/audio_helper_test.rb)

It took me a while to understand why these [specs](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/test/audio_helper_test.rb#L27-L39) were failing. when I realised that this [block](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/lib/lotus/helpers/audio_helper.rb#L14) is being evaluated in the `HtmlBuilder` context when I was expecting it to call the methods _(source and track)_ I wrote in the `AudioHleper`. Yes, I know, I'm not very smart. :)

So I had to start to work in a new approach for the audio helper, which's an [audio_builder](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/lib/lotus/helpers/auido_helper/audio_builder.rb) like the work done with form builder, which I'm not 100% sure it's the best way to do that.

**TL;DR**
- *Problem*, I want the execute/evaluate [block](https://github.com/klebervirgilio/helpers/blob/asset_helper_audio/lib/lotus/helpers/audio_helper.rb#L14) into a context where the source and track methods are the ones I've written.
- *Possible Solution*, the only way I can see to do that is to work on something similar to form_builder.

Sorry, I have a poor English. Does that make sense?
Thanks for the help!

**Comment by klebervirgilio** [comment]
Any thoughts?

**Comment by joneslee85** [comment]
Sorry for late reply, I'll have a look 2moro when I'm back to my computer

**Comment by jodosha** [comment]
@klebervirgilio Are you still interested to implement this helper?

**Please notice that this shouldn't be implemented here, but in `lotus-assets`. Please check https://github.com/lotus/assets/blob/master/lib/lotus/assets/helpers.rb and `video` helper for reference.**

Please let me know. Thanks!

**Comment by klebervirgilio** [comment]
@jodosha yes, I am. I was waiting for the feedback in the code I've already written.


> Hi @joneslee85, thanks for reaching out.
>
>I believe I have already done most of the work, you can check it out here.
>
>My initial work were basically this humble module for which I've written these specs
>
>It took me a while to understand why these specs were failing. when I realised that this block is being evaluated into the HtmlBuilder context when I was expecting it to call the methods (source and track) that I've written in the AudioHleper. Yes, I know, I'm not very smart. :)
>
>So I had to start to work in a new approach for the audio helper, which's an audio_builder like the work done with form builder, which I don't fully understand and I'm not 100% sure it's the best approach as ?well.
>
>TL;DR
>
>Problem, I want to execute/evaluate block into a context where the source and track methods are the ones I've written.
>Possible Solution, the only way I can see to do that is to work on something similar to form_builder.
>Sorry, I have a poor English. Does that make sense?
>Thanks for the help

**Comment by jodosha** [comment]
@klebervirgilio Hey :smile: 

Quoting myself:
> Please check https://github.com/lotus/assets/blob/master/lib/lotus/assets/helpers.rb and `video` helper for reference.

You implemented that feature in `lotus/lotus` repository (aka `lotusrb` gem), but that should be in `lotus/assets` repository (aka `lotus-assets` gem). I would prefer to have the same `video` helper implementation and tests because their pretty similar helpers.

Thank you! :+1:

**Comment by klebervirgilio** [comment]
@jodosha great! thanks for the support will work on that.

**Comment by klebervirgilio** [comment]
https://github.com/lotus/assets/pull/29
Simple, video-tag-like and with tests covering all the requirements described here. Hopefully, everything is good. Let me know if you guys need more help.

**Comment by jodosha** [comment]
@klebervirgilio It's perfect :+1: Thank you

**Comment by jodosha** [comment]
Closing as implemented by https://github.com/lotus/assets/pull/29
